flowmachine/flowmachine/features/subscriber/home_location_monthly.py

```diff
@@ -0,0 +1,197 @@
+    (choose X=15 here), then choose this as their home location. If not, but it was their last location on more than Y
+     days this month (Y<X; choose Y=10 here) AND it is the same as their “reference location” (in our case reference
+     location is modal location and last location on more than X days last month), then choose this as their home
+     location. Otherwise set their home location to “unlocatable” (a special value, not NULL)."""
+
+    def __init__(
+        self,
+        window_start: Union[str, datetime],
+        window_stop: Union[str, datetime],
+        agg_unit: AnySpatialUnit,
+        home_this_month: int,
+        home_last_month: int,
+        ref_location: Union["HomeLocationMonthly", None] = None,
+        events_tables=None,
+        modal_lookback=40,
+        active_days=4,
+        interval=7,
+    ):
+
+        self.window_start = window_start
+        self.window_stop = window_stop
+        self.ref_location = ref_location
+        self.agg_unit = agg_unit
+        self.home_this_month = home_last_month
+        self.home_last_month = home_this_month
+        self.modal_lookback = modal_lookback
+
+        self.active_subs = ActiveSubscribers(
+            start_date=self.window_start,
+            end_date=self.window_stop,
+            active_days=active_days,
+            interval=interval,
+            events_tables=events_tables,
+        )
+
+        if self._window_start >= self._window_stop - timedelta(days=modal_lookback):
+            full_range_start = self._window_stop - timedelta(days=modal_lookback)
+        else:
+            full_range_start = self._window_start
+        full_range_stop = self._window_stop
+
+        all_locations = {
+            day: LastLocation(
+                start=day.strftime("%Y-%m-%d"),
+                stop=(day + timedelta(days=1)).strftime("%Y-%m-%d"),
+                spatial_unit=self.agg_unit,
+                subscriber_subset=self.active_subs,
+                table=events_tables,
+            )
+            for day in rrule(DAILY, dtstart=full_range_start, until=full_range_stop)
+        }
+        modal_days = [
+            query
+            for (day, query) in all_locations.items()
+            if full_range_start <= day <= full_range_stop
+        ]
+        self.last_locations = [
+            query
+            for (day, query) in all_locations.items()
+            if self._window_start <= day <= self._window_stop
+        ]
+        self.modal_locations = ModalLocation(*modal_days)
+
+        # self.daily_location_frequency = LocationVisits(
+        #     self.last_locations  # Ask Jono about this
+        # )
+
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        cols = ["subscriber", "location", "month"]
+        return cols
+
+    @property
+    def window_start(self):
+        return self._window_start.strftime("%Y-%m-%d")
+
+    @window_start.setter
+    def window_start(self, value):
```

Comment:
Thinking about it, I'm not overwhelmingly opposed to using getters and setters for this stuff, but I'd rather they were defined once in the base class.

Reply:
I could have sworn I saw a mixin for doing this that already existed, but when I went looking for it I couldn't find it. Think we need a new one?

Reply:
I'm going to leave them in for now, as it's useful to be able to have the dates as Datetime internally but exposed as a string for compatability with the rest of Flowmachine - agree that they need to be defined once, though.